File: mobile-android/app/src/main/kotlin/com/mortitech/treasuryflow/feature/auth/data/repository/AuthRepository.kt
package com.mortitech.treasuryflow.feature.auth.data.repository

import com.mortitech.treasuryflow.core.common.AppResult
import com.mortitech.treasuryflow.core.common.appResultOf
import com.mortitech.treasuryflow.core.datastore.TokenStorage
import com.mortitech.treasuryflow.core.model.User
import com.mortitech.treasuryflow.core.network.ApiService
import com.mortitech.treasuryflow.core.session.SessionManager
import com.mortitech.treasuryflow.feature.auth.data.remote.dto.*
import com.mortitech.treasuryflow.feature.auth.domain.AuthProvider
import kotlinx.coroutines.flow.firstOrNull
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val apiService: ApiService,
    private val tokenStorage: TokenStorage,
    private val sessionManager: SessionManager,
) {

    suspend fun oauthCallback(provider: AuthProvider, code: String, redirectUri: String): AppResult<User> =
        appResultOf {
            val response = apiService.oauthCallback(OAuthCallbackRequest(provider, code, redirectUri))
            tokenStorage.saveTokens(response.accessToken, response.refreshToken)
            response.user.toDomain()
        }


    suspend fun register(email: String, password: String, displayName: String): AppResult<User> =
        appResultOf {
            val response = apiService.register(RegisterRequest(email, password, displayName))
            tokenStorage.saveTokens(response.accessToken, response.refreshToken)
            response.user.toDomain()
        }

    suspend fun login(email: String, password: String): AppResult<User> =
        appResultOf {
            val response = apiService.login(LoginRequest(email, password))
            tokenStorage.saveTokens(response.accessToken, response.refreshToken)
            response.user.toDomain()
        }


    suspend fun refreshToken(): AppResult<User> =
        appResultOf {
            val refreshToken = tokenStorage.refreshToken.firstOrNull()
                ?: throw IllegalStateException("No refresh token available")
            val response = apiService.refreshToken(RefreshTokenRequest(refreshToken))
            tokenStorage.saveTokens(response.accessToken, response.refreshToken)
            response.user.toDomain()
        }

    suspend fun getCurrentUser(): AppResult<User> =
        appResultOf {
            apiService.getCurrentUser().toDomain()
        }

    suspend fun logout() {
        sessionManager.onLogout()
    }
}

File: mobile-ios/treasuryflow/Data/Repository/AuthRepository.swift
import Foundation

protocol AuthRepository: Sendable {
    func oauthCallback(provider: String, code: String, redirectUri: String) async throws -> AuthResponse

    func register(email: String, password: String, displayName: String) async throws -> AuthResponse
    func login(email: String, password: String) async throws -> AuthResponse

    func refreshToken() async throws -> AuthResponse
    func getCurrentUser() async throws -> UserResponse
    func logout() async
}

final class AuthRepositoryImpl: AuthRepository, Sendable {
    private let apiClient: APIClient
    private let tokenStorage: TokenStorage

    init(apiClient: APIClient, tokenStorage: TokenStorage) {
        self.apiClient = apiClient
        self.tokenStorage = tokenStorage
    }

    func oauthCallback(provider: String, code: String, redirectUri: String) async throws -> AuthResponse {
        let body = OAuthCallbackRequest(provider: provider, code: code, redirectUri: redirectUri)
        let response: AuthResponse = try await apiClient.request(.oauthCallback(), body: body)
        await tokenStorage.saveTokens(accessToken: response.accessToken, refreshToken: response.refreshToken)
        return response
    }


    func register(email: String, password: String, displayName: String) async throws -> AuthResponse {
        let body = RegisterRequest(email: email, password: password, displayName: displayName)
        let response: AuthResponse = try await apiClient.request(.register(), body: body)
        await tokenStorage.saveTokens(accessToken: response.accessToken, refreshToken: response.refreshToken)
        return response
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        let body = LoginRequest(email: email, password: password)
        let response: AuthResponse = try await apiClient.request(.login(), body: body)
        await tokenStorage.saveTokens(accessToken: response.accessToken, refreshToken: response.refreshToken)
        return response
    }


    func refreshToken() async throws -> AuthResponse {
        guard let refreshToken = await tokenStorage.getRefreshToken() else {
            throw APIError.unauthorized
        }
        let body = RefreshTokenRequest(refreshToken: refreshToken)
        let response: AuthResponse = try await apiClient.request(.refreshToken(), body: body)
        await tokenStorage.saveTokens(accessToken: response.accessToken, refreshToken: response.refreshToken)
        return response
    }

    func getCurrentUser() async throws -> UserResponse {
        try await apiClient.request(.me())
    }

    func logout() async {
        await tokenStorage.clearTokens()
    }
}
